x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2867 @@
+
+This means that **no changes will be needed in this page.**
+
+#### Via the Stack Management > Rules UI
+
+The Stack Management Rules UI provides a unified view across the Elastic stack, showing not only security rules, but also other types such as observability, ML, etc.
+
+All rules listed here can be modified in the following ways:
+
+- enabling and disabling
+- snoozing/unsnoozing or scheduling/unscheduling snooze
+- updating their API keys
+- deleting them
+
+All of these actions are currently applicable for both prebuilt security rules as well as custom security rules, and they use the Alerting Framework endpoints of:
+
+- `POST /alerting/rules/_bulk_edit`
+- `PATCH /alerting/rules/_bulk_delete`
+
+This means that **no changes will be needed in this page.**
+
+## List of things to fix (create tickets)
+
+1. Validation of EQL queries incorrectly fails: valid EQL queries that come with Prebuilt Rules are rejected by the EQL validator when trying to edit a rule. See [link](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/public/detections/components/rules/step_define_rule/schema.tsx#L186).
+2. Custom Query field overflows the viewport and cannot be completely visualized/edited when the query is too long, i.e. occupies too many lines. This is often the case from Prebuilt Rules. The editable field also does not allow scrolling. See example:
+   ![](2023-12-01-14-13-07.png)
+
+
+
+## Upgrading Prebuilt Rules
+
+### Changes to upgrade `_review` endpoint
+
+The algorithm for calculating rule version diffs' core structure is [fully implemented](https://github.com/elastic/kibana/pull/148392) and operational. However, the current method for calculating field differences is basic: it currently simply compares field values across versions and flags a conflict if the base version, the current version and the target versions differ from one another. In this conflict scenario, the `mergedVersion` of the field proposed by the algorithm always equals the `targetVersion`.
+
+We propose developing more adaptable diff algorithms tailored to specific rule fields or rule field types. These more specific algorithms aim to minimize conflicts and automate merging of changes, doing the best effort to keep the intended customizations applied by the user, as well as the updates proposed by Elastic. Hopefully, the user will be able to simply review the proposal and accept it.
+
+#### Concrete field diff algorithms by type
+
+Depending on the specific field or type of field we might want to apply a specific merging algorithm when conflicts arise. Let's propose different types.
+
+##### String fields
+
+For string fields, in case of three-way conflicts, we will make a best effort to reconcile any modifications done by the user and any updates proposed by Elastic.
+
+Examples: `name`, `description`, `setup`, `note` (Investigation guide)
+
+  <table>
```

Comment:
I echo @maximpn's comment here, it looks like overengineering to me too.

First, I would treat single-line fields like `name` (especially if they are keywords like `severity`) and multi-line text fields like `description` and `setup` differently.

Then, for single-line fields we could use the existing simple diff algorithm:

- if `base !== current !== target`, we say it's a **conflict** and use the `current` version as the merged one
- if `base === target && base !== current`, we set `merged = current` without a conflict
- if `base === current && base !== target`, we set `merged = target` without a conflict

For multi-line text or markdown fields I think we should implement a git-like merge algorithm where changes in different areas of the text would be mergeable, but changes to the same line would render a conflict.

Finally, a couple of thoughts about single-line string fields:

- Merging keywords (especially enums) should never be done, because it could generate an invalid value.
- Changes to a rule's name might indicate some semantical changes to the whole rule (e.g. related to its source data, query/filters logic, etc). So, if both Elastic and user changed the name of the same rule, this could be an indication that the two versions of these rule are not compatible with each other, and the changes need to be reviewed by the user. Generating a conflict in this case would help the user to pay attention to all the changes.

Reply:
Splitted string diff algorithms section into two tables, simplified them and proposed a library we can use for multi-line string diffs.

Reply:
@jpdjere The updated text looks great, thank you!